### ISSUE TITLE
MovingPathPoint derives from Station but it should not

### ISSUE THREAD

**Issue by chrisdembia** [opensim-org/opensim-core#1179]
MovingPathPoint derives from Station even though it is not fixed to a Body. Lots of things could conceivably be done with Stations that should not be done with other Points, like defining constraints.

This inheritance from Station was introduced by #946  and also came up in #990 and #1177. In #990, @aseth1 attempted to prevent misuse by setting the location property to NaN for MovingPathPoints.

**Comment by tkuchida** [comment]
> prevent misuse by setting the location property to NaN

Setting the `location` property to `NaN` did not prevent misuse; see #1104.

**Comment by chrisdembia** [comment]
How about "to prevent silent misuse."? It would have been *worse* if you *did* see the moving path points but they weren't moving.

**Comment by tkuchida** [comment]
> It would have been *worse* if you *did* see the moving path points but they weren't moving.

Actually, it would have been easier to spot if `location` were being set to `Vec3(0)` or `Vec3(huge_number)`; I didn't immediately notice there were missing line segments in gait10dof18musc.

What is the proposal? Should MovingPathPoint derive from Point? If so, ConditionalPathPoint would be the only descendant of PathPoint—should PathPoint be absorbed into ConditionalPathPoint?

**Comment by chrisdembia** [comment]
> Actually, it would have been easier to spot if location were being set to Vec3(0) or Vec3(huge_number); I didn't immediately notice there were missing line segments in gait10dof18musc.

Does the same hold true for calculations that use the location, like moment arm or path length? I think `Vec3(huge_number)` might be nearly as good of a red flag as `NaN`, but `Vec3(0)` would not be.

> What is the proposal? Should MovingPathPoint derive from Point?

Something like that. @aseth1 doesn't want special PathPoint classes at all. Just Point classes that the muscles use.

**Comment by tkuchida** [comment]
> Does the same hold true for calculations that use the location, like moment arm or path length?

Good point. Perhaps `Vec3(huge_number)` then, or just throw an exception if `MovingPathPoint::get_location()` is called.

> @aseth1 doesn't want special PathPoint classes at all. Just Point/Station classes that the muscles use.

Okay, I won't go after this, then. Thanks for the heads-up.

**Comment by chrisdembia** [comment]
Yeah I don't think the fix is straightforward.

**Comment by jenhicks** [comment]
@aseth1 Will move PathPoint out of Station.

**Comment by jenhicks** [comment]
Could throw exception when user asks for location of the moving path point as an intermediate step.

**Comment by tkuchida** [comment]
> Could throw exception when user asks for location of the moving path point as an intermediate step.

Investigated overriding `get_location()`, `upd_location()`, and `set_location()` (which are inherited from Station) to throw exceptions. It doesn't look like this would be a "quick fix" since `set_location()` is being called somewhere via `Model::initSystem()`. We could print a warning, but these warnings would appear when `initSystem()` is called even if the user has done nothing wrong.

**Comment by chrisdembia** [comment]
I don't think `get_location()` etc are virtual. Anywho, if that isn't a quick fix, then it's probably easier to just change the inheritance hierarchy.

**Comment by aymanhab** [comment]
I do think that the right fix is changing the inheritance hierarchy indeed and not to override the method. Also the GUI accesses the Properties directly not through the custom interface get/set so that wouldn't have helped much. We probably need to size the scope of the change and its side-effects this late in the game before we commit.

**Comment by aseth1** [comment]
IMHO the correct solution is to remove the `PathPoint` class altogether since it is an inappropriate base for representing both a stationary point and a moving point. The problem was not a new one, only exacerbated by the fact that we now have a clearer delineation between stationary (`Station`) and other points. With the introduction of `Point` a path can be a simply composed of a collection of Points. If PathPoints have some other intrinsic property (e.g. they can provide their partial derivative with respect to the `PhysicalFrame` they are connected too, then we can add that to the base `Point` which should be straightforward. 

`PathPoint` was left mostly in tact because of the following methods:
```cpp
    const double& getLocationCoord(int aXYZ) const;

    // A variant that uses basic types for use by GUI
    void setLocationCoord(int aXYZ, double aValue);

    void setLocation( const SimTK::State& s, int aCoordIndex, double aLocation);

#ifndef SWIG
    // A variant that uses basic types for use by GUI
    void setLocation( const SimTK::State& s, double pt[]);
#endif
    void setBody(const PhysicalFrame& body);

    void changeBodyPreserveLocation(const SimTK::State& s,
                                    const PhysicalFrame& body);


    virtual void scale(const SimTK::Vec3& scaleFactors);

    virtual const WrapObject* getWrapObject() const { return NULL; }

    virtual bool isActive(const SimTK::State& s) const { return true; }

    // get the relative velocity of the path point with respect to the body
    // it is connected to.
    virtual void getVelocity(const SimTK::State& s, SimTK::Vec3& velocity) {
        velocity.setToZero();
    }
    // get the partial of the point location w.r.t. to the coordinates (Q)
    // it is dependent on.
    virtual SimTK::Vec3 getdPointdQ(const SimTK::State& s) const
        { return SimTK::Vec3(0); }

    virtual void updateGeometry() {}

    static void deletePathPoint(PathPoint* aPoint) { if (aPoint) delete aPoint; }
```
In my view,  these methods are not "standard" (e.g. having a static method to delete a point) and confusing like `getVelocity` (that is relative to the body)  and we could adhere to and extend the `Point` interface to accomplish the same ends. @aymanhab which if these are essential for editing and manipulation of paths in the GUI?

**Comment by aymanhab** [comment]
@aseth1 working on a fix that changes the class hierarchy, main concern is interaction with wrapping

**Comment by jenhicks** [comment]
What is the time estimate to complete @aseth1 ?

**Comment by aseth1** [comment]
Working on changes discussed today and it should be ready to test soon. I am hoping to make a PR before I leave today.

**Comment by aseth1** [comment]
Fixed by #1595
